### ISSUE TITLE
Refactor API for load/store OP and read-only attachments

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#2131]
Spawned from https://github.com/gpuweb/gpuweb/issues/2062#issuecomment-910669436

Our current API for declaring operations and read-only properties is not ideal:
```js
    required (GPULoadOp or float) depthLoadValue;
    required GPUStoreOp depthStoreOp;
    boolean depthReadOnly = false;
```

Problems with it:
  1. If my attachment doesn't have a depth component, I'm still forced to specify both load and store ops. The store op has to be "store" even though there can be no store performed. Same for stencil.
  2. If the depth aspect is read-only, I'm still forced to specify the store operation as "store". Same for stencil.

Diagram for understanding the data and validation:
![WebGPUAttachmentOps](https://user-images.githubusercontent.com/107301/134092420-48e01b66-2d31-43c4-96d6-dcd9f6e4e5a1.png)

**Comment by kvark** [comment]
Strawman proposal A
```js
dictionary GPUAttachmentOps {
  required GPULoadOp loadOp;
  required GPUStoreOp storeOp;
};

enum GPUReadOnlyAttachment {
  "read-only"
};

dictionary GPURenderPassDepthStencilAttachment {
    required GPUTextureView view;

    float depthClearValue = 0.0;
    (GPUReadOnlyAttachment or GPUAttachmentOps) depth; //nullable

    GPUStencilValue stencilClearValue = 0;
    (GPUReadOnlyAttachment or GPUAttachmentOps) stencil; //nullable
};
```

If an aspect isn't used, user just doesn't specify it, and it's null.
If it's used, then it's either "read-only", or the structure with load/store ops.

**Comment by kainino0x** [comment]
Sounds like roughly where I would have gone. GPULoadOp would have to gain a "clear" variant, but otherwise seems ok. For consistency with color attachments, how about:

```js
dictionary GPUDepthAttachmentOps {
  required (GPULoadOp or float) loadOp;
  required GPUStoreOp storeOp;
};

dictionary GPUStencilAttachmentOps {
  required (GPULoadOp or GPUStencilValue) loadOp;
  required GPUStoreOp storeOp;

enum GPUReadOnlyAttachment {
  "read-only"
};

dictionary GPURenderPassDepthStencilAttachment {
    required GPUTextureView view;

    (GPUReadOnlyAttachment or GPUDepthAttachmentOps) depth; // optional
    (GPUReadOnlyAttachment or GPUStencilAttachmentOpsOps) stencil; // optional
};
```

**Comment by kainino0x** [comment]
See #2132 which may affect exactly how we design this. E.g.: initially disallow {depth,stencil} from being undefined if view doesn't have a {depth,stencil} aspect; then, later:
1.  Make undefined mean "aspect is unused".
2. Add a new enum variant to allow `depth: "unused"`/`stencil: "unused"`.

**Comment by Kangz** [comment]
While it's nice to be able to specify that the render pass doesn't write some aspects of the attachment, I'm a bit worried by the depth of the type tree: it represents the possibilities exactly, but is conceptually more complicated. What about saying that if `depthReadOnly` is `true`, then `depthLoadValue` and `depthStoreOp` are ignored? Ditto if the aspect isn't present: we just ignore the value of the member for that aspect. (this would mean the only required member of `GPURenderPassDepthStencilAttachment` is `view`).

**Comment by kainino0x** [comment]
editors: Here's a new idea:

```
enum GPURenderPassDepthStencilAttachmentMode {
    "clear-render-discard",
    "clear-render-store",
    "load-render-discard",
    "load-render-store",
    "read-only",
}

dictionary GPURenderPassDepthStencilAttachment {
    required GPUTextureView view;

    float depthClearValue = 0.0; // ... or 1.0?
    required GPURenderPassDepthStencilAttachmentMode depthMode;

    GPUStencilValue stencilClearValue = 0;
    required GPURenderPassDepthStencilAttachmentMode stencilMode;
};
```

Applying this design to color attachments for consistency (or just separating `loadValue` back out the way it was a long time ago) would also obsolete #1377, because there would be a "clear" mode and a separate clear value.

```
enum GPURenderPassAttachmentMode {
    "clear-render-discard",
    "clear-render-store",
    "load-render-discard",
    "load-render-store",
}

dictionary GPURenderPassColorAttachment {
    required GPUTextureView view;
    GPUTextureView resolveTarget;

    GPUColor loadValue; // default to [0,0,0,0]?
    required GPURenderPassAttachmentMode mode;
};
```

**Comment by kvark** [comment]
And for completeness, here is another choice for us:
```js
enum GPURenderPassAttachmentLoadOp {
  "clear",
  "load",
}

enum GPURenderPassAttachmentMutability {
  "store",
  "discard",
  "read-only",
}

dictionary GPURenderPassDepthStencilAttachment {
    required GPUTextureView view;

    float depthClearValue = 0.0; // ... or 1.0?
    required GPURenderPassAttachmentLoadOp depthLoadOp;
    required GPURenderPassAttachmentMutability depthMutability;

    GPUStencilValue stencilClearValue = 0;
    required GPURenderPassAttachmentLoadOp stencilLoadOp;
    required GPURenderPassAttachmentMutability stencilMutability;
};
```

The aspect of appeal in this one is that it keeps the load/store distinction as they are separate bits of state, while still integrating the read-only-ness.
What this doesn't solve, however, is that the user still needs to specify stencil operations even if there is no stencil.
For this, the `GPURenderPassDepthStencilAttachmentMode` approach is more tractable, since we can just add an "unused" variant to it as a default (or just make it nullable?).

**Comment by Kangz** [comment]
Can this be closed now that #2387 is landed?
